modyn/storage/internal/file_wrapper/csv_file_wrapper.py

```diff
@@ -0,0 +1,198 @@
+                    raise ValueError("The label must be an integer")
+
+        if len(set(number_of_columns)) != 1:
+            raise ValueError(
+                "Some rows have different width. " f"This is the number of columns row by row {number_of_columns}"
+            )
+
+    def get_sample(self, index: int) -> bytes:
+        samples = self._filter_rows_samples([index])
+
+        if len(samples) != 1:
+            raise IndexError("Invalid index")
+
+        return samples[0]
+
+    def get_samples(self, start: int, end: int) -> list[bytes]:
+        indices = list(range(start, end))
+        return self.get_samples_from_indices(indices)
+
+    def get_samples_from_indices(self, indices: list) -> list[bytes]:
+        samples = self._filter_rows_samples(indices)
+
+        if len(samples) != len(indices):
+            raise IndexError("At least one index is invalid.")
+
+        return samples
+
+    def get_label(self, index: int) -> Optional[int]:
+        if self.label_index is None:
+            return None
+
+        labels = self._filter_rows_labels([index])
+
+        if len(labels) != 1:
+            raise IndexError("Invalid index.")
+
+        return labels[0]
+
+    def get_all_labels(self) -> list[Optional[int]]:
+        reader = self._get_csv_reader()
+
+        labels: list[Optional[int]] = []
+
+        if self.label_index is None:
+            return [None] * self.get_number_of_samples()
+
+        for row in reader:
+            # labels are integer in modyn
+            int_label = int(row[self.label_index])
+            labels.append(int_label)
+
+        return labels
+
+    def get_number_of_samples(self) -> int:
+        reader = self._get_csv_reader()
+
+        return sum(1 for _ in reader)
+
+    def _get_csv_reader(self) -> Iterator:
+        """
+        Receives the bytes from the file_system_wrapper and creates a csv.reader out of it.
+        Returns:
+            csv.reader
+        """
+        data_file = self.filesystem_wrapper.get(self.file_path)
+
+        # Convert bytes content to a string
+        data_file_str = data_file.decode(self.encoding)
+
+        lines = data_file_str.split("\n")
+
+        # Create a CSV reader
+        reader = csv.reader(lines, delimiter=self.separator)
+
+        # skip the header if required
+        if self.ignore_first_line:
+            next(reader)
+
+        return reader
+
+    def _filter_rows_samples(self, indices: list[int]) -> list[bytes]:
+        """
+        Filters the selected rows and removes the label column
+        Args:
+            indices: list of rows that must be kept
+
+        Returns:
+            list of byte-encoded rows
+
+        """
+        reader = self._get_csv_reader()
+
+        # Iterate over the rows and keep the selected ones
+        filtered_rows = []
+        for i, row in enumerate(reader):
+            if i in indices:
```

Comment:
i think converting indices to a set is helpful to speed this up (also everywhere where you do these checks)

Reply:
In this version yes, but now if I want to keep the same order I think it is more convenient to have a list